File: Java/ExtensibleCommands/src/test/java/org/extensiblecommands/GenericCyclicCommandTest.java
package org.extensiblecommands;

import org.junit.*;
import org.junit.rules.TestName;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static java.lang.Thread.sleep;

public class GenericCyclicCommandTest {
    @Rule
    public TestName name = new TestName();

    @BeforeClass
    public static void setUpClass() {
        Setup.InitLog();
    }

    @Before
    public void setUpTest() {
        Logger.log(Logger.LogLevel.Info,
                "----------------------------------------------------------------------------------------------------------");
        Logger.log(Logger.LogLevel.Info,
                String.format("Starting Test %s:%s", this.getClass().getName(), name.getMethodName()));
    }

    @Test
    public void constructionTest() {
        var command = new GenericCyclicCommand<Integer>(SimpleCommand.NullCommand, Arrays.asList(2));
        Assert.assertEquals(command.getCoreCommand(), SimpleCommand.NullCommand);
        Assert.assertEquals(command.getName(), "Generic Cyclic");

        command = new GenericCyclicCommand<Integer>(SimpleCommand.NullCommand, Arrays.asList(2), "MyCommand");
        Assert.assertEquals(command.getCoreCommand(), SimpleCommand.NullCommand);
        Assert.assertEquals(command.getName(), "MyCommand");

        // Malformed cases
        boolean exceptionCaught = false;
        try {
            new GenericCyclicCommand<Integer>(null, null);
        }
        catch (Exception e) {
            if (e.getMessage().contains("is NULL"))
                exceptionCaught = true;
        }

        Assert.assertTrue(exceptionCaught);

        exceptionCaught = false;
        try {
            new GenericCyclicCommand<Integer>(SimpleCommand.NullCommand, null, "");
        }
        catch (Exception e) {
            if (e.getMessage().contains("is NULL"))
                exceptionCaught = true;
        }
        Assert.assertTrue(exceptionCaught);
    }

    @Test
    public void runOkTest() throws Exception {
        var command = createCyclicCommand();
        Setup.runAndWaitForNormalCompletion(command);
        Assert.assertEquals(3, command.getCurrentCycle());
        Assert.assertEquals(30, (int)command.getCurrentElement());  // Enumerator should iterate until the end of the collection and be equal to last element
    }

    @Test
    public void runErrorTest() throws Exception {
        var command = createErrorCyclicCommand();

        Setup.runAndWaitForFailure(command);
        Assert.assertEquals(command.getCoreCommand().getState(), State.Failed);
        Assert.assertEquals(1, command.getCurrentCycle());
        Assert.assertEquals(10, (int)command.getCurrentElement());  // Enumerator should be at 1st element

        Assert.assertEquals(Setup.TestErrorCode, command.getException().getId());
        Assert.assertEquals(Setup.TestErrorDescription, command.getException().getText());
    }

    @Test
    public void pauseResumeTest() throws Exception {
        var cyclicCommand = createPauseAbortCyclicCommand(true);

        ExecutionDelegate assertAfterPause = () -> {
            Assert.assertEquals(State.Executing, cyclicCommand.getCoreCommand().getState());
            Assert.assertEquals(1, cyclicCommand.getCurrentCycle());
        };

        Setup.pauseAndResume(cyclicCommand, assertAfterPause);

        Assert.assertEquals(State.Completed, cyclicCommand.getState());
        Assert.assertEquals(State.Completed, cyclicCommand.getCoreCommand().getState());
        Assert.assertEquals(3, cyclicCommand.getCurrentCycle());
    }

    @Test
    public void pauseAbortTest() throws Exception {
        var cyclicCommand = createPauseAbortCyclicCommand(true);

        ExecutionDelegate assertAfterPause = () -> {
            Assert.assertEquals(State.Executing, cyclicCommand.getCoreCommand().getState());
            Assert.assertEquals(1, cyclicCommand.getCurrentCycle());
        };

        Setup.pauseAndAbort(cyclicCommand, assertAfterPause);

        Assert.assertEquals(State.Aborted, cyclicCommand.getState());
        Assert.assertEquals(State.Aborted, cyclicCommand.getCoreCommand().getState());
        Assert.assertEquals(1, cyclicCommand.getCurrentCycle());
    }

    @Test
    public void abortTest() throws Exception {
        var cyclicCommand = createPauseAbortCyclicCommand(false);

        Setup.runAndWaitForAbort(cyclicCommand);

        Assert.assertEquals(State.Aborted, cyclicCommand.getCoreCommand().getState());
        Assert.assertEquals(1, cyclicCommand.getCurrentCycle());
    }

    @Test
    public void externalAbortTest() throws Exception {
        var command = createCyclicCommand();

        Setup.runAndAbort(command);

        Assert.assertEquals(State.Aborted, command.getState());
        Assert.assertEquals(State.Aborted, command.getCoreCommand().getState());
    }

    @Test
    public void retrieveSubCommandsTest() {
        var command = createCyclicCommand();

        Assert.assertEquals(2, command.getCoreCommand().getDescendants().spliterator().getExactSizeIfKnown());
        Assert.assertEquals(3, command.getDescendants().spliterator().getExactSizeIfKnown());
        Assert.assertEquals(2, command.getCoreCommand().getChildren().spliterator().getExactSizeIfKnown());
        Assert.assertEquals(1, command.getChildren().spliterator().getExactSizeIfKnown());
    }

    @Test
    public void runInputOutputTest() throws Exception {
        var command1 = new SimpleCommandIO<String, Integer>(input -> input.length(), "Test");

        var list = new ArrayList<Integer>() { { add(10); add(20); add(30); }};
        var coreCommand = new SequentialCommand();
        var cyclicCommand = new GenericCyclicCommand<>(coreCommand, list);

        coreCommand.add(new SimpleCommand(() -> { int k = cyclicCommand.getCurrentElement(); }))
            .add(new SimpleCommand(() -> sleep((int)(0.6 * Setup.ThreadLatencyDelayMsec)), "Sleep"))
            .add(command1);

        command1.setInput("input");
        cyclicCommand.run();
    }

    //----------------------------------------------------------------------------------------------------------------------

    private GenericCyclicCommand<Integer> createCyclicCommand() {
        List<Integer> list = Arrays.asList(10, 20, 30);

        var coreCommand = new SequentialCommand();
        var command = new GenericCyclicCommand<>(coreCommand, list);

        coreCommand.add(new SimpleCommand(() -> { int k = command.getCurrentElement(); }))
            .add(new SimpleCommand(() -> sleep((Setup.ThreadLatencyDelayMsec)), "Sleep"));
        return command;
    }

    private GenericCyclicCommand<Integer> createErrorCyclicCommand() {
        List<Integer> list = Arrays.asList(10, 20, 30);

        var coreCommand = new SequentialCommand();
        var command = new GenericCyclicCommand<>(coreCommand, list);

        coreCommand.add(new SimpleCommand(() -> { int k = command.getCurrentElement(); }))
            .add(new SimpleCommand(() -> { throw new ExtensibleCommandsException(Setup.TestErrorCode, Setup.TestErrorDescription); }, "Sleep"));
        return command;
    }

    private GenericCyclicCommand<Integer> createPauseAbortCyclicCommand(boolean pause) {
        List<Integer> list = Arrays.asList(10, 20, 30);

        var coreCommand = new SequentialCommand("Core 2-step sequential command")
            .add(new SimpleCommand(() -> sleep((int)(0.3 * Setup.ThreadLatencyDelayMsec)), "Sleep"));

        var cyclicCommand = new GenericCyclicCommand<>(coreCommand, list, "Cyclic test command");

        if (pause)
            coreCommand.add(new SimpleCommand(() -> {
                // Only pause during the first cycle
                if (cyclicCommand.getCurrentCycle() == 1)
                    cyclicCommand.pause();
            }));
        else
            coreCommand.add(new SimpleCommand(() -> {
                // Only abort during the first cycle
                if (cyclicCommand.getCurrentCycle() == 1)
                    cyclicCommand.abort();
                }));
        return cyclicCommand;
    }
}

File: CSharp/ExtensibleCommands/ExtensibleCommandsUnitTests/GenericCyclicCommandTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Threading;

using ExtensibleCommands;

namespace ExtensibleCommandsUnitTest
{
    /// <summary>
    /// Summary description for GenericCyclicCommandTest
    /// </summary>
    [TestClass]
    public class GenericCyclicCommandTest
    {
        private TestContext testContextInstance;

        /// <summary>
        ///Gets or sets the test context which provides
        ///information about and functionality for the current test run.
        ///</summary>
        public TestContext TestContext
        {
            get
            {
                return testContextInstance;
            }
            set
            {
                testContextInstance = value;
            }
        }

        #region Additional test attributes
        //
        // You can use the following additional attributes as you write your tests:
        //
        // Use ClassInitialize to run code before running the first test in the class
        [ClassInitialize]
        public static void MyClassInitialize(TestContext testContext)
        {
        }
        //
        // Use ClassCleanup to run code after all tests in a class have run
        [TestInitialize]
        public void MyTestInitialize()
        {
            Logger.Log(Logger.LogLevel.Info,
                "----------------------------------------------------------------------------------------------------------");
            Logger.Log(Logger.LogLevel.Info,
                string.Format("Starting Test {0}:{1}", GetType().Name, testContextInstance.TestName));
        }
        //
        // Use TestInitialize to run code before running each test 
        // [TestInitialize()]
        // public void MyTestInitialize() { }
        //
        // Use TestCleanup to run code after each test has run
        // [TestCleanup()]
        // public void MyTestCleanup() { }
        //
        #endregion

        [TestMethod()]
        public void ConstructionTest()
        {
            var command = new GenericCyclicCommand<int>(SimpleCommand.NullCommand, new List<int> { 2 });
            Assert.AreEqual(command.CoreCommand, SimpleCommand.NullCommand);
            Assert.AreEqual(command.Name, "Generic Cyclic");

            command = new GenericCyclicCommand<int>(SimpleCommand.NullCommand, new List<int> { 2 }, "MyCommand");
            Assert.AreEqual(command.CoreCommand, SimpleCommand.NullCommand);
            Assert.AreEqual(command.Name, "MyCommand");

            // Malformed cases
            bool exceptionCaught = false;
            try
            {
                new GenericCyclicCommand<int>(null, null);
            }
            catch (Exception e)
            {
                if (e.Message.Contains("is NULL"))
                    exceptionCaught = true;
            }
            Assert.IsTrue(exceptionCaught);

            exceptionCaught = false;
            try
            {
                new GenericCyclicCommand<int>(SimpleCommand.NullCommand, null, "");
            }
            catch (Exception e)
            {
                if (e.Message.Contains("is NULL"))
                    exceptionCaught = true;
            }
            Assert.IsTrue(exceptionCaught);
        }

        [TestMethod()]
        public void RunOKTest()
        {
            var command = CreateCyclicCommand();
            Setup.RunAndWaitForNormalCompletion(command);
            Assert.AreEqual(5, command.CurrentCycle);
            Assert.AreEqual(50, command.CurrentElement);  // Enumerator should iterate until the end of the collection and be equal to last element
        }

        [TestMethod()]
        public void RunErrorTest()
        {
            var command = CreateErrorCyclicCommand();

            Setup.RunAndWaitForFailure(command);
            Assert.AreEqual(command.CoreCommand.CurrentState, State.Failed);
            Assert.AreEqual(1, command.CurrentCycle);
            Assert.AreEqual(10, command.CurrentElement);  // Enumerator should be at 1st element

            Assert.AreEqual(Setup.TestErrorCode, command.Exception.ID);
            Assert.AreEqual(Setup.TestErrorDescription, command.Exception.Text);
        }

        [TestMethod()]
        public void PauseResumeTest()
        {
            var cyclicCommand = CreatePauseAbortCyclicCommand(true);

            var assertAfterPause = new System.Action(() =>
            {
                Assert.AreEqual(State.Executing, cyclicCommand.CoreCommand.CurrentState);
                Assert.AreEqual(1, cyclicCommand.CurrentCycle);
            });

            Setup.PauseAndResume(cyclicCommand, assertAfterPause);

            Assert.AreEqual(State.Completed, cyclicCommand.CurrentState);
            Assert.AreEqual(State.Completed, cyclicCommand.CoreCommand.CurrentState);
            Assert.AreEqual(3, cyclicCommand.CurrentCycle);
        }

        [TestMethod()]
        public void PauseAbortTest()
        {
            var cyclicCommand = CreatePauseAbortCyclicCommand(true);

            var assertAfterPause = new System.Action(() =>
            {
                Assert.AreEqual(State.Executing, cyclicCommand.CoreCommand.CurrentState);
                Assert.AreEqual(1, cyclicCommand.CurrentCycle);
            });

            Setup.PauseAndAbort(cyclicCommand, assertAfterPause);

            Assert.AreEqual(State.Aborted, cyclicCommand.CurrentState);
            Assert.AreEqual(State.Aborted, cyclicCommand.CoreCommand.CurrentState);
            Assert.AreEqual(1, cyclicCommand.CurrentCycle);
        }

        [TestMethod()]
        public void AbortTest()
        {
            var cyclicCommand = CreatePauseAbortCyclicCommand(false);

            Setup.RunAndWaitForAbort(cyclicCommand);

            Assert.AreEqual(State.Aborted, cyclicCommand.CoreCommand.CurrentState);
            Assert.AreEqual(1, cyclicCommand.CurrentCycle);
        }

        [TestMethod()]
        public void ExternalAbortTest()
        {
            var command = CreateCyclicCommand();

            Setup.RunAndAbort(command);

            Assert.AreEqual(State.Aborted, command.CurrentState);
            Assert.AreEqual(State.Aborted, command.CoreCommand.CurrentState);
        }

        [TestMethod()]
        public void RetrieveSubCommandsTest()
        {
            var command = CreateCyclicCommand();

            Assert.AreEqual(2, command.CoreCommand.Descendants.Count());
            Assert.AreEqual(3, command.Descendants.Count());
            Assert.AreEqual(2, command.CoreCommand.Children.Count());
            Assert.AreEqual(1, command.Children.Count());
        }

        [TestMethod()]
        public void RunInputOutputTest()
        {
            var command = new SimpleCommandIO<string, int>(input => input.Length, "Test");

            var list = new List<int> { 10, 20, 30 };
            var coreCommand = new SequentialCommand();
            var cyclicCommand = new GenericCyclicCommand<int>(coreCommand, list);

            coreCommand.Add(new SimpleCommand(() => { int k = cyclicCommand.CurrentElement; }))
                .Add(new SimpleCommand(() => Thread.Sleep((int)(0.6 * Setup.ThreadLatencyDelayMsec)), "Sleep"))
                .Add(command);

            command.Input = "input";
            cyclicCommand.Run();
        }

        //----------------------------------------------------------------------------------------------------------------------

        private GenericCyclicCommand<int> CreateCyclicCommand()
        {
            var list = new List<int> {10, 20, 30, 40, 50};
            var coreCommand = new SequentialCommand();
            var command = new GenericCyclicCommand<int>(coreCommand, list);

            coreCommand.Add(new SimpleCommand(() => { int k = command.CurrentElement; }))
                .Add(new SimpleCommand(() => Thread.Sleep(Setup.ThreadLatencyDelayMsec), "Sleep"));
            return command;
        }

        private GenericCyclicCommand<int> CreateErrorCyclicCommand()
        {
            var list = new List<int> { 10, 20, 30 };
            var coreCommand = new SequentialCommand();
            var command = new GenericCyclicCommand<int>(coreCommand, list);

            coreCommand.Add(new SimpleCommand(() => { int k = command.CurrentElement; }))
                .Add(new SimpleCommand(() => { throw new ExtensibleCommandsException(Setup.TestErrorCode, Setup.TestErrorDescription); }, "Sleep"));
            return command;
        }

        private GenericCyclicCommand<int> CreatePauseAbortCyclicCommand(bool pause)
        {
            var list = new List<int> { 10, 20, 30 };
            GenericCyclicCommand<int> cyclicCommand = null;
            var coreCommand = new SequentialCommand("Core 2-step sequential command");
            coreCommand.Add(new SimpleCommand(() => Thread.Sleep((int)(0.3 * Setup.ThreadLatencyDelayMsec)), "Sleep"));
            if (pause)
                coreCommand.Add(new SimpleCommand(() =>
                {
                    // Only pause during the first cycle
                    if (cyclicCommand.CurrentCycle == 1)
                        cyclicCommand.Pause();
                }));
            else
                coreCommand.Add(new SimpleCommand(() =>
                {
                    // Only abort during the first cycle
                    if (cyclicCommand.CurrentCycle == 1)
                        cyclicCommand.Abort();
                }));
            cyclicCommand = new GenericCyclicCommand<int>(coreCommand, list, "Cyclic test command");
            return cyclicCommand;
        }
    }
}
